### ISSUE TITLE
@vanilla-extract/vite-plugin loads css files multiple times in Storybook when using theme

### ISSUE THREAD

**Issue by SamBernaerdtOntoforce** [vanilla-extract-css/vanilla-extract#1488]
### Describe the bug

**Minimal Problem:**
When using themes, vite loads a new `(filename).css.ts.vanilla.css?t=1728553686586` file for every other file that uses itin the development build of storybook.

<img width="1339" alt="image" src="https://github.com/user-attachments/assets/ddcbaeb7-c1a5-4e35-88ea-3b509cc220cd">

**Expected:**
Each .css.ts file should only be loaded once (until it changes and causes a hot reload)

**Problem:**
This might not be an issue on the small reproducible project, but on a larger project, where multiple theme files are composited, each of the theme files is loaded for each component using a theme value. This causes so many file loads that it freezes storybook.

<img width="1084" alt="image" src="https://github.com/user-attachments/assets/a39c84e8-180a-4ec7-8210-229ee3e276fd">

### Reproduction

https://github.com/SamBernaerdtOntoforce/vanilla-extract-theme-and-storybook

### System Info

```shell
System:
    OS: macOS 14.7
    CPU: (12) arm64 Apple M2 Pro
    Memory: 279.81 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.7.3 - ~/.nvm/versions/node/v21.7.3/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v21.7.3/bin/yarn
    npm: 10.5.0 - ~/.nvm/versions/node/v21.7.3/bin/npm
    pnpm: 8.8.0 - ~/.nvm/versions/node/v21.7.3/bin/pnpm
  Browsers:
    Chrome: 129.0.6668.100
    Edge: 129.0.2792.79
    Safari: 18.0
  npmPackages:
    @vanilla-extract/css: ^1.16.0 => 1.16.0
    @vanilla-extract/vite-plugin: ^4.0.16 => 4.0.16
    vite: ^5.4.8 => 5.4.8
```

### Used Package Manager

yarn

### Validations

- [X] Check that there isn't [already an issue](https://github.com/vanilla-extract-css/vanilla-extract/issues) that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by askoufis** [comment]
Thanks for the reproduction! It definitely looks like there could be an optimization here. This doesn't seem related specifically to Storybook as I can reproduce the issue using just a vite dev server.

It seems that the issue starts once HMR has kicked in at least once. Different files that share the same dependency (e.g. `shared.css.ts`) end up with different timestamped imports of that file, causing both of them to be loaded, even though only 1 is necessary.

I'm also seeing some inconsistency in how many `.vanilla.css` files are loaded. Occasionally each file is loaded only once:
<img width="864" alt="image" src="https://github.com/user-attachments/assets/1a7e225d-1e5f-4807-a335-11d2dc02253e">

However most of the time the same file is loaded a few times (but with different timestamps):
<img width="905" alt="image" src="https://github.com/user-attachments/assets/31b518ad-cd36-402d-a7af-bbaae61f03c0">

This could be a result of how vite loads modules during development, and maybe there's nothing we can do to change it. I'll need to dig into it some more.

**Comment by hartz89** [comment]
I am experiencing this exact same issue.  I'm working in a `pnpm` monorepo and just migrated our storybook to the vite builder.  It's very problematic and a total performance killer in dev mode in a large project with a lot of VE CSS.  I'm investigating, but I'm  new to the internals of Vanilla Extract and Vite so any guidance/pointers would be appreciated.

**Comment by hartz89** [comment]
I have opened PR #1591 to address this issue with a new plugin option.  Please let me know if you have feedback on the approach.  Initial builds + HMR were working as expected in my local testing.

**Comment by oliveryasuna** [comment]
> Also worth noting that, even without this PR, the following config change drastically improved startup times because it minized dependency re-optimization and related HMR churn:
> 
> export default defineConfig({
>   plugins: [vanillaExtractPlugin()],
>   optimizeDeps: {
>     include: [
>       '@vanilla-extract/recipes/createRuntimeFn',
>       '@vanilla-extract/sprinkles/createRuntimeSprinkles',
>     ],
>   }
> });
> 
> I wonder if the plugin should do that by default, or if it's better to let consumers opt-in. In any event, would it be a worthwhile addition to the Vite plugin docs?

This also fixes HMR issues in Storybook with Vite.

**Comment by askoufis** [comment]
I believe I've addressed at least part of this issue in [this branch](https://github.com/vanilla-extract-css/vanilla-extract/compare/vite-hmr-fix?expand=1). You can try out this fix in `@vanilla-extract/vite-plugin@5.1.1-vite-hmr-fix-20250706080631`. It should at least prevent unnecessary loading of the same module multiple times with a different HMR timestamp.

@hartz89 As I mentioned in [my discord reply to your message](https://discord.com/channels/885877446098964512/1384994419924668476/1385436055674814524), I'm unsure if a file cache is the correct solution here. The compiler is already caching generated CSS, and ultimately the correct thing to do when a leaf node module is invalidate is to invalidate the module tree upwards. Importer modules can't have any knowledge of what has changed in the leaf node, so the best approach to get accurate HMR is to re-build everything.

I will investigate your `optimizeDeps` solution too.

@SamBernaerdtOntoforce Please try this out in a larger project if you have one. IMO splitting out your theme tokens in such a granular way is part of the problem here, though without understanding your use case I can't say for certain whether there's a better way to structure them. Regardless, hopefully this fix should help your situation a bit.

**Comment by kompot** [comment]
@askoufis tried `@vanilla-extract/vite-plugin@5.1.1-vite-hmr-fix-20250706080631` on our project

we have heavily sprinkles based CSS of atomic nature and almost every component's `css.ts` imports some sprinkle from a pretty complex sprinkles.css.ts file

basically nothing has changed — cold start is very rough, I see like 100 requests to a sprinkles file (see screenshot) 

we have everything vanilla extract related added to `optmizeDeps.include` and that helps a bit

```
include: [
        '@vanilla-extract/css-utils',
        '@vanilla-extract/css',
        '@vanilla-extract/css/fileScope',
        '@vanilla-extract/dynamic',
        '@vanilla-extract/recipes',
        '@vanilla-extract/recipes/createRuntimeFn',
        '@vanilla-extract/sprinkles',
        '@vanilla-extract/sprinkles/createRuntimeSprinkles',
],
```

<img width="300px" alt="Image" src="https://github.com/user-attachments/assets/90ec7bb7-0f94-4456-86cd-203fbf65e91e" />

after first initial rought sequence everything is kind of ok (usually no more than 2 requests to sprinkles, one of them with a timestamp)

<img width="568" height="54" alt="Image" src="https://github.com/user-attachments/assets/7c3ee25e-135f-4301-8ab5-2e4e99655ce5" />

but this branch also introduced these warnings to console, you might be interested

```
11:35:33 AM [vite] (client) Pre-transform error: Internal server error: Soft-invalidated module "SOME_REACT_COMPONENT.tsx" should not have existing transform result
11:35:34 AM [vite] (client) Pre-transform error: Internal server error: Soft-invalidated module "SOME_VANILLA_EXTRACT_STYLESHEET.css.ts" should not have existing transform result
11:35:34 AM [vite] (client) Pre-transform error: Internal server error: Soft-invalidated module "SOME_PLAIN_TYPESCRIPT_FILE.ts" should not have existing transform result
```

it's absent in main branch

happy to test possible new fixes

UPD
have tested HMR of a very deeply nested sprinkle and seems like situation is different (and  better): in main branch I had to actually restart dev server to see any changes in this sprinkle, in this branch it's HMRed but multiple requests you see on the screenshot above are happening all the time, on each change

**Comment by askoufis** [comment]
@kompot Thanks for the feedback, this is very useful. Maybe not invalidating the first time a module is transformed will improve cold start? I'll have to test that out.

Those soft-invalidation errors are strange - I wouldn't have expected them to show up as a result of the fix.

Would something like 100 identical react components all importing the same sprinkles file be a good enough reproduction of your situation?

**Comment by kompot** [comment]
> Maybe not invalidating the first time a module is transformed will improve cold start? I'll have to test that out

not sure how the whole mechanism actually works but maybe last update time should be tracked and if multiple modules have requested the same vanilla extract module then they should wait for the same, only once created, transformation promise 

judging only by my screenshot I can guess that `t` query string parameter means `time the module was requested`

> Would something like 100 identical react components all importing the same sprinkles file be a good enough reproduction of your situation?

well, looks like it might be enough 

not sure what causes this request hell though, I mean maybe the fact that our sprinkles are pretty complex and are transformed kinda slow that leads to the next component requesting raw sprinkles file instead of already transformed result (again, no idea how the internals work, just guessing)

considering that `it's all plain typescript` you could probably add artifical delays in `css.ts` file itself to test slow transformations?

**Comment by kompot** [comment]
@askoufis I have run vite with `DEBUG=vite:*` env param and observe that vanilla extract takes a lot time indeed and transforms each file at least twice.

```
vite:transform 1582.12ms /src/default/components/LicenseMsg/LicenseMsg.instance.css.ts +1ms
  vite:load 0.19ms [plugin] /src/default/components/LicenseMsg/LicenseMsg.instance.css.ts.vanilla.css +3ms
  vite:hmr [self-accepts] src/default/components/LicenseMsg/LicenseMsg.instance.css.ts.vanilla.css +3ms
  vite:import-analysis 0.04ms [0 imports rewritten] src/default/components/LicenseMsg/LicenseMsg.instance.css.ts.vanilla.css +0ms
  vite:transform 0.12ms /src/default/components/LicenseMsg/LicenseMsg.instance.css.ts.vanilla.css +0ms
  vite:import-analysis 0.72ms [6 imports rewritten] src/default/components/LicenseMsg/LicenseMsg.instance.css.ts +0ms
  vite:transform 1578.77ms /src/default/components/LicenseMsg/LicenseMsg.instance.css.ts +1ms



  vite:load 0.61ms [fs] src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +1ms
  vite:import-analysis 1.04ms [5 imports rewritten] src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +3ms
  vite:transform 1.96ms src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +2ms
  vite:import-analysis 0.53ms [3 imports rewritten] src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +10ms
  vite:transform 1593.97ms /src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +10ms
  vite:load 0.42ms [plugin] /src/default/styling/rules/layout.css.ts.vanilla.css +12ms
  vite:load 0.43ms [plugin] /src/default/styling/sprinkles.css.ts.vanilla.css +0ms
  vite:load 0.49ms [plugin] /src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts.vanilla.css +0ms
  vite:import-analysis 0.96ms [19 imports rewritten] src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +0ms
  vite:transform 1591.28ms /src/default/components/limitsAndFees/LimitsAndFeesList.instance.css.ts +0ms

```

as you see it tried to transform `LimitsAndFeesList.instance.css.ts` and `LicenseMsg.instance.css.ts` twice spending 1500ms each time and that is on MacBook Pro M2 Pro, not a particularly slow machine. Also these lines are just one after another (not different recompilations).

**Comment by christhegrand** [comment]
When will a fix for this issue be merged?

**Comment by askoufis** [comment]
@christhegrand @kompot @hartz89 @oliveryasuna @SamBernaerdtOntoforce Please try out this snapshot version: `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250826075423`. In my testing it looks like it fixes the duplicate loading issue, and reduces (potentially eliminates) unnecessary HMR invalidations.

**Comment by kompot** [comment]
Confirming 0 extra requests on cold start (after `rm -rf ./node_modules/.vite`). Also no `ts` in querystring.

before (with 5.1.1)

<img width="488" height="220" alt="Image" src="https://github.com/user-attachments/assets/2918b255-9939-43af-b48f-e736fb5fde22" />

after 

<img width="666" height="103" alt="Image" src="https://github.com/user-attachments/assets/de6510ef-82bd-4d43-ab80-5c478316bbe8" />

But observed behavior was broken unfortunately. 

With `5.1.1` if I change something deeply nested I get full page reload and changes are applied (with tens of requests to sprinkles.css.ts?ts=XXX). With `5.1.2-vite-hmr-fix-20250826075423` I also get full page reload but nothing changes.

**Comment by askoufis** [comment]
@kompot Ah, I thought I'd tested the deeply nested use case. I guess it doesn't work. Would it be possible for you to make a minimal reproduction of your specific situation? That would help a lot.

**Comment by kompot** [comment]
@askoufis Here's minimal repo to reproduce
https://github.com/kompot/vanilla-extract-vite-plugin-hmr-testcase

```shell
pnpm install
pnpm start
```

It will start with `5.1.2-vite-hmr-fix-20250826075423`. Open root page and you'll see 3 headers.

<img width="358" height="240" alt="Image" src="https://github.com/user-attachments/assets/3f63e600-b176-4588-bd10-ab63416fe993" />

updating this `border-radius` should HMR red's rectangular border radius
https://github.com/kompot/vanilla-extract-vite-plugin-hmr-testcase/blob/51ce9a4acbccacda28a8c95ab0eabd9799ae54a4/src/styling/rules/border.css.ts#L3

but it does not (hard reload does not help)

with 5.1.1 everything works as it should.

**Comment by askoufis** [comment]
@kompot Thank you for that. It has helped me narrow down the issue to a tricky edge case. I'll report back if I come up with a solution.

**Comment by ramiroazar** [comment]
This was really slowing down the Vite dev server in our case, but can also confirm the `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250826075423` snapshot resolves this for us.

**Comment by askoufis** [comment]
@ramiroazar `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250911065547` is a more recent snapshot that fixes some invalidation edge cases.

@kompot The newer snapshot _almost_ completely eliminates unnecessary invalidations in your test repo, however for some reason `App.tsx` is invalidated twice. However, it's better than the current state, so I might make a PR for it regardless.

**Comment by ramiroazar** [comment]
Thanks, @askoufis, `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250911065547` is working great, 

I agree that this is not a Storybook specific issue by the way.

**Comment by ryanone** [comment]
FWIW I've been trying this out for several days now and have no longer been experiencing any issues. Thanks for publishing this snapshot version.

**Comment by sambernaerdtphysibel** [comment]
> [@christhegrand](https://github.com/christhegrand) [@kompot](https://github.com/kompot) [@hartz89](https://github.com/hartz89) [@oliveryasuna](https://github.com/oliveryasuna) [@SamBernaerdtOntoforce](https://github.com/SamBernaerdtOntoforce) Please try out this snapshot version: `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250826075423`. In my testing it looks like it fixes the duplicate loading issue, and reduces (potentially eliminates) unnecessary HMR invalidations.

Apologies for the late reply.

I've just updated to this snapshot and it looks like this resolved the issue. Thanks a lot for the fix!

**Comment by hartz89** [comment]
@askoufis I just gave `@vanilla-extract/vite-plugin@5.1.2-vite-hmr-fix-20250911065547` a spin and it seems to be working great, vastly improved the initial page load and HMR on my vite apps!

If there's any hesitation around the stability of the change, would it make sense to control the behavior with a configuration option / "future flag" for the initial release?  Would be happy to use it in dev for a few months to sus out any potential corner cases, but I can't really do that with a snapshot version.

**Comment by sparta-dan** [comment]
5.1.2-vite-hmr-fix-20250911065547 improves things, still not perfect, but i see it did not make 5.1.3, will this get an official patch?

**Comment by askoufis** [comment]
> 5.1.2-vite-hmr-fix-20250911065547 improves things, still not perfect, but i see it did not make 5.1.3, will this get an official patch?

@sparta-dan What behaviour are you seeing that you don't expect? Any examples/reproductions would be extremely helpful so I find more edge cases.

**Comment by sparta-dan** [comment]
@askoufis we are on a vite app in an NX monorepo. we are having multiple requests for each `css.ts` file too. dozens per file in the monorepo, causing our page reload time to be ~40s+. the `5.1.2-vite-hmr-fix-20250911065547` version reduces the duplicates greatly, and gets us to 2-3s page reloads. It still has some duplication, but it is a great improvement. 5.1.3 did not include the hotfix, and still causes us issues.

We are using this hotfix version for our SaaS but would like to use a formal version. One of my colleagues is digging into the issue and hotfix to further optimise and avoid any duplication... hopefully we will have a PR for you to help the cause.

Many thanks,

<img width="339" height="462" alt="Image" src="https://github.com/user-attachments/assets/f6c6e724-cd76-4723-ae69-e57b2ed8c049" />

**Comment by askoufis** [comment]
The latest version of the vite plugin released some fixes for HMR that should help reduce extra invalidations. https://github.com/vanilla-extract-css/vanilla-extract/releases/tag/%40vanilla-extract%2Fvite-plugin%405.2.1

It might not fix _all_ extra invalidations, but should hopefully improve things. If anyone is still having issues with multiple HMR invalidations, _**please raise a separate issue with a minimal reproduction**_. I'd prefer to have new issue(s) with smaller reproductions separate to this issue which was originally more focused on Storybook.

Thank you all for your feedback and patience with this issue.
